File: frontend/src/lib/api/billing.ts
import apiClient from './client';

export interface BillingStatus {
  is_active: boolean;
  is_trialing: boolean;
  plan_type: string;
  provider: string | null;
  trial_ends_at: string | null;
  current_period_end: string | null;
  subscription: {
    id: string;
    provider: string | null;
    plan_type: string;
    status: string;
    trial_ends_at: string | null;
    current_period_start: string | null;
    current_period_end: string | null;
    canceled_at: string | null;
    created_at: string;
    updated_at: string;
  } | null;
}

export interface PaymentProvider {
  id: string;
  name: string;
  label: string;
}

export interface CheckoutResponse {
  checkout_url: string;
  session_id: string | null;
  provider: string;
}

export const billingApi = {
  /** Get available payment providers configured on the backend. */
  providers: async (): Promise<{ providers: PaymentProvider[] }> => {
    const response = await apiClient.get('/billing/providers');
    return response.data;
  },

  /** Get the current user's billing/subscription status. */
  status: async (): Promise<BillingStatus> => {
    const response = await apiClient.get('/billing/status');
    return response.data;
  },

  /** Create a checkout session to subscribe via the given provider. */
  checkout: async (
    provider: string,
    plan: string = 'pro',
    successUrl?: string,
    cancelUrl?: string,
  ): Promise<CheckoutResponse> => {
    const response = await apiClient.post('/billing/checkout', {
      provider,
      plan,
      success_url: successUrl,
      cancel_url: cancelUrl,
    });
    return response.data;
  },

  /** Create a customer portal session for billing management. */
  portal: async (): Promise<{ portal_url: string; provider: string }> => {
    const response = await apiClient.post('/billing/portal');
    return response.data;
  },
};

export default billingApi;

File: backend/app/schemas/billing.py
"""Billing and subscription schemas."""
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class SubscriptionResponse(BaseModel):
    """Schema returned when querying a user's current subscription state."""
    id: UUID
    provider: Optional[str] = None
    provider_customer_id: Optional[str] = None
    plan_type: str = "free"
    status: str = "trialing"
    trial_ends_at: Optional[datetime] = None
    current_period_start: Optional[datetime] = None
    current_period_end: Optional[datetime] = None
    canceled_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CheckoutRequest(BaseModel):
    """Schema for initiating a checkout session."""
    provider: str  # "stripe" or "paddle"
    plan: str = "pro"
    success_url: Optional[str] = None
    cancel_url: Optional[str] = None


class CheckoutResponse(BaseModel):
    """Schema returned after creating a checkout session."""
    checkout_url: str
    session_id: Optional[str] = None
    provider: str


class CustomerPortalResponse(BaseModel):
    """Schema returned with a link to the billing management portal."""
    portal_url: str
    provider: str


class BillingStatusResponse(BaseModel):
    """High-level billing overview for the frontend UI."""
    is_active: bool
    is_trialing: bool
    plan_type: str
    provider: Optional[str] = None
    trial_ends_at: Optional[datetime] = None
    current_period_end: Optional[datetime] = None
    subscription: Optional[SubscriptionResponse] = None
